### PR TITLE
Implement mutiple inputs and update mleap dependency version with 0.14.0

### PR DESCRIPTION
**Pull request by jinpengqi**
*This PR is linked to https://github.com/aws/sagemaker-sparkml-serving-container/pull/16 which is already merged.*
All code change in this pr is included in above one. So close this pr now.

*Description of changes:*
1. Accepting multiple json inputs 
2. Upgrade the mleap to 0.14.0 and sparkml-serving to 2.4

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 1]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
         }
     }
 
+    /**
+     * Implements the invocations POST API for application/jsonlines input
+     *
+     * @param jsonLines, lines of json values
+     * @param accept, accept parameter from request
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:156]: Accept indicates the content types that the http method is able to understand. Can you change the documentation to indicate that? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 2]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+     * @return ResponseEntity with body as the expected payload JSON & proper statuscode based on the input
+     */
+    @RequestMapping(path = "/invocations", method = POST, consumes = AdditionalMediaType.APPLICATION_JSONLINES_VALUE)
+    public ResponseEntity<String> transformRequestJsonLines(@RequestBody final byte[] jsonLines,
+                                                            @RequestHeader(value = HttpHeaders.ACCEPT, required = false) final String accept) {
+        if (jsonLines == null) {
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:162]: Does this capture the empty case as well or just the null case? You could potentially have something like this as input: new byte[0]; You might want to catch that.

Comment (mahithsuresh) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:162]: And fix the log statement to "input passed to the request is null or empty."

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 3]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+            LOG.error("Input passed to the request is empty");
+            return ResponseEntity.noContent().build();
+        }
+        try {
+            final String acceptVal = this.retrieveAndVerifyAccept(accept);
+            final DataSchema schema = this.retrieveAndVerifySchema(null, mapper);
+            final String jsonStringLine = new String(jsonLines);
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:168]: Might want to name it jsonLinesAsString just to be accurate.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 4]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+            for (ResponseEntity<String> response:responseList) {
+                HttpStatus statuscode = response.getStatusCode();
+                headers = response.getHeaders();
+                bodyList.add(Lists.newArrayList(response.getBody()));
+            }
+
+            return ResponseEntity.ok().headers(headers).body(bodyList.toString());
+        } catch (final Exception ex) {
+            LOG.error("Error in processing current request", ex);
+            return ResponseEntity.badRequest().body(ex.getMessage());
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:198]: Will ex.getMessage() output the stack trace as well? How about you just pass in ex to the body? You might want to test that.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 5]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+            return ResponseEntity.noContent().build();
+        }
+        try {
+            final String acceptVal = this.retrieveAndVerifyAccept(accept);
+            final DataSchema schema = this.retrieveAndVerifySchema(null, mapper);
+            final String jsonStringLine = new String(jsonLines);
+
+            // Map list of inputs to DataList object
+            final SageMakerDataListObject sro = mapper.readValue(jsonStringLine, SageMakerDataListObject.class);
+            List<List<Object>> inputDatas = sro.getData();
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:173]: You don't need inputDatas and inputDatasSize. Just use it directly.

 for e.g. for (int idx = 0; idx < sro.getData().size(); idx++)...

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 6]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+            List<ResponseEntity<String>> responseList = Lists.newArrayList();
+
+            // Process each input separately and add response to a list
+            final int inputDatasSize = inputDatas.size();
+            for (int idx = 0; idx < inputDatasSize; ++idx) {
+                ResponseEntity<String> response = this.processInputData(inputDatas.get(idx), schema, acceptVal);
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:179]: Same as above - don't need response. => responseList.add(this.processInputData(...))

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 7]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+                responseList.add(response);
+            }
+
+            // Merge response body to a new output response
+            List<List<String>> bodyList = Lists.newArrayList();
+            HttpHeaders headers = null;
+            //combine body in responseList
+            for (ResponseEntity<String> response:responseList) {
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:191]: Leave a whitespace after ":"

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 8]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+                HttpStatus statuscode = response.getStatusCode();
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java]: unncessary. Remove and use directly.

Comment (mahithsuresh) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java]: Btw - looks like we are not using it at all.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 9]
```diff
@@ -148,6 +150,54 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+            final DataSchema schema = this.retrieveAndVerifySchema(null, mapper);
+            final String jsonStringLine = new String(jsonLines);
+
+            // Map list of inputs to DataList object
+            final SageMakerDataListObject sro = mapper.readValue(jsonStringLine, SageMakerDataListObject.class);
+            List<List<Object>> inputDatas = sro.getData();
+            List<ResponseEntity<String>> responseList = Lists.newArrayList();
+
+            // Process each input separately and add response to a list
+            final int inputDatasSize = inputDatas.size();
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:177]: I guess you could potentially combine this block and the next into a separate method called processInput and then just add the output from that to the final response list. That way - it might be more readable?

File: src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java [hunk 10]
```diff
@@ -230,6 +230,42 @@ public void testListValueMLeapThrowsExceptionCsvApi() {
...
+        PowerMockito
+                .when(ScalaUtils.getJavaObjectIteratorFromArrayRow(Mockito.any(ArrayRow.class), Mockito.anyString()))
+                .thenThrow(new RuntimeException("input data is not valid"));
+        final ResponseEntity<String> output = controller.transformRequestJsonLines("{\"data\":[[1,2.0,\"TEST1\"], [2,3.0,\"TEST\"]]}".getBytes(), "text/csv");
+        Assert.assertEquals(output.getStatusCode(), HttpStatus.BAD_REQUEST);
+        Assert.assertEquals(output.getBody(), "input data is not valid");
+    }
+
+    @Test
+    public void testJsonLinesApiWithNullInput() {
```
Comment (mahithsuresh) [inline src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java:264]: Add one more test for empty input (new byte[])

File: src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java [hunk 11]
```diff
@@ -230,6 +230,42 @@ public void testListValueMLeapThrowsExceptionCsvApi() {
...
+        PowerMockito
+                .when(ScalaUtils.getJavaObjectIteratorFromArrayRow(Mockito.any(ArrayRow.class), Mockito.anyString()))
+                .thenReturn(outputResponseForFirstInput.iterator())
+                .thenReturn(outputResponseForSecondInput.iterator());
+        final ResponseEntity<String> output = controller.transformRequestJsonLines("{\"data\":[[1,2.0,\"TEST1\"], [2,3.0,\"TEST\"]]}".getBytes(), "text/csv");
+        Assert.assertEquals(expectOutput, output.getBody());
+    }
+
+    @Test
+    public void testJsonLinesApiWithListInputThorwsException() {
```
Comment (mahithsuresh) [inline src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java]: Just log and make sure that it is outputting the stack trace.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 1 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:156]: Modified

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 2 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:162]: When we passed byte[0], it will return bad request instead of no content for response.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 3 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:168]: Modified

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 4 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:198]: ex is Exception type but here responseEntity requires the body as String so we may use getMessage here.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 5 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:173]: Modified

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 7 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:191]: Modified

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 8 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java]: Yes, we changed already.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java (hunk 9 above)
Comment (jinpengqi) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:177]: Yes, we may create a new method called processInputDataForMutiplelinesInput since we already had a method called processInputData.

File: src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java (hunk 10 above)
Comment (jinpengqi) [reply src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java:264]: Tested
1. For new byte[] input, the output will be <400 BAD_REQUEST Bad Request> .
2. For null input, the output will be <204 NO_CONTENT No Content>.
Should we expect them to be the same error throwing or this should be fine?

File: src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java (hunk 11 above)
Comment (jinpengqi) [reply src/test/java/com/amazonaws/sagemaker/controller/ServingControllerTest.java]: Here I think it will just output the exception message to us instead of the whole stack trace which is incorporated in old code pattern.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 12]
```diff
@@ -148,6 +150,31 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+     * Implements the invocations POST API for application/jsonlines input
+     *
+     * @param jsonLines, lines of json values
+     * @param accept, indicates the content types that the http method is able to understand
+     * @return ResponseEntity with body as the expected payload JSON & proper statuscode based on the input
+     */
+    @RequestMapping(path = "/invocations", method = POST, consumes = AdditionalMediaType.APPLICATION_JSONLINES_VALUE)
+    public ResponseEntity<String> transformRequestJsonLines(@RequestBody final byte[] jsonLines,
+                                                            @RequestHeader(value = HttpHeaders.ACCEPT, required = false) final String accept) {
+        final byte[] emptyInput = new byte[0];
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:163]: You can just do jsonLines.length == 0 right? Doesnt that work?

Comment (mahithsuresh) [reply src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:163]: And don't we need to check for null as well? I'd recommend explicitly handling the null input case and returning a bad request instead of just letting it flow and raising an exception.

File: src/main/java/com/amazonaws/sagemaker/controller/ServingController.java [hunk 13]
```diff
@@ -148,6 +150,31 @@ public ResponseEntity returnBatchExecutionParameter() throws JsonProcessingExcep
...
+        if (Arrays.equals(emptyInput, jsonLines)) {
+            LOG.error("Input passed to the request is empty");
+            return ResponseEntity.noContent().build();
+        }
+        try {
+            final String jsonLinesAsString = new String(jsonLines);
```
Comment (mahithsuresh) [inline src/main/java/com/amazonaws/sagemaker/controller/ServingController.java:169]: My bad - I didn't realize earlier - you probably don't need this variable. You can use it directly?

File: src/test/java/com/amazonaws/sagemaker/dto/SageMakerRequestListObjectTest.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+                Lists.newArrayList(39.0, 25.0)));
+    }
+
+    @Test
+    public void testSageMakerRequestObjectCreation() throws IOException {
+        String inputJson = IOUtils.toString(this.getClass().getResourceAsStream("basic_input_schema.json"), "UTF-8");
+        DataSchema schema = mapper.readValue(inputJson, DataSchema.class);
+        SageMakerRequestListObject sro = new SageMakerRequestListObject(schema, listOfListInputForBasicInput);
+        Assert.assertEquals(sro.getSchema().getInput().size(), 3);
+        Assert.assertEquals(sro.getSchema().getInput().get(0).getName(), "name_1");
```
Comment (mahithsuresh) [inline src/test/java/com/amazonaws/sagemaker/dto/SageMakerRequestListObjectTest.java:55]: There is some redundant code here across tests with all the assert statements - could you combine the verifications into a single method and reuse that in all the tests?
